### ISSUE TITLE
key_options are no longer set with authorized_key

### ISSUE THREAD

**Issue by jurgenhaas** [ansible/ansible#5486]
When running this task

```
- name: 'Subversion | Install SSH user public keys'
  authorized_key: user=sshsvn
                  key="{{lookup('file', inventory_dir + '/files/certs/' + item + '.svn.pub')}}"
                  key_options='command="/usr/local/bin/svnserve -t --tunnel-user={{item}}",no-port-forwarding,no-agent-forwarding,no-X11-forwarding,no-pty'
  with_items: svn_user
```

then the key options are no longer added to the ~/.ssh/authorized_keys file on the remote host anymore. This used to be working prior to version 1.4 final but is no longer working since. Tested against 1.4.1, 1.4.3 and 1.5dev/HEAD.

To reproduce, simply run the above task where the variable svn_user is a list of usernames as strings and the file "item.svn.pub" is a SSH RSA public key file.

**Comment by mpdehaan** [comment]
So I believe @jctanner's comment was to look on the development branch.

Looking at the code, it looks like key options are appended to the key to see if the key is there or not, but that it would not modify an existing key, instead adding a duplicate. (this differs from the bug report, though that may also be a problem that they are not added, though code review doesn't point at this)

I would say that the test should be for the key itself, and then the key options field would modify an existing key match.   Thus you could use the module to modify key options -- which isn't happening.

We should test to ensure the key is in place, and that key options can be modified without adding a duplicate.

**Comment by jurgenhaas** [comment]
Let me add that I have also deleted the entire file authorized_keys and tried again. Only the keys got added without the key options.

**Comment by drybjed** [comment]
Confirming the issue. When `key_options=no-port-forwarding,no-agent-forwarding` is defined, options are added to the key, but with `key_options=no-port-forwarding,no-agent-forwarding,command=\"git-shell -c \\\"$SSH_ORIGINAL_COMMAND\\\"\"` only key is added.

**Comment by jctanner** [comment]
@jurgenhaas and @drybjed Please test with this patch:

```
diff --git a/library/system/authorized_key b/library/system/authorized_key
index 7626a9a..74ebe5c 100644
--- a/library/system/authorized_key
+++ b/library/system/authorized_key
@@ -251,7 +251,11 @@ def parsekey(module, raw_key):
         options = key_parts[0]

     # parse the options (if any)
-    options = parseoptions(module, options)
+    if module.params['key_options'] is not None:
+        # https://github.com/ansible/ansible/issues/5486
+        options = parseoptions(module, module.params['key_options'])
+    else:
+        options = parseoptions(module, options)

     # get key after the type index
     key = key_parts[(type_index + 1)]
```

It fixes the example in the description without breaking any of my existing tests. Please let me know of any use cases that do not work for you so that I can add tests for them.

I'm guessing I might have to iterate further or refactor the module more, but I hope this patch takes care of it.

**Comment by jurgenhaas** [comment]
@drybjed thanks for looking into this. I have mixed test results here.

The good thing: it is now including the options into the authorized_keys file.

The bad thing: from my use case above it does not properly resolve the {{item}} for --tunner-user in the options. One of the values in with_items is used for all lines in the resulting file.

**Comment by gwillem** [comment]
Thanks @jctanner, working great for me! Minor remark: from semantic perspective, perhaps it's cleaner to concatenate the explicit options from the yaml with the "implicit" options from the key file, instead of silently discarding the latter.

**Comment by drybjed** [comment]
For me, with the path applied, `authorized_key` behaves completely randomly - sometimes it applies key_options as intended (on one key), sometimes it puts key_options from one key into all keys, other times there's no key_options at all. Maybe it has something to do with default(""), is there a way to have authorized_key interpret key_options="" as having no key options? Here's my task that configures authorized keys:

```
- name: Configure authorized keys for users
  authorized_key: key='{{ item.key }}' manage_dir=no key_options='{{ item.options | default("") }}'
                  path=/etc/ssh/authorized_keys/{{ item.user }} user={{ item.user }}
                  state={{ item.state | default('present') }}
  with_flattened:
    - sshd_authorized_keys_list
    - sshd_authorized_keys_group_list
    - sshd_authorized_keys_host_list
  when: item is defined and item and item.key is defined and item.user is defined
```

And example key list:

```
sshd_authorized_keys_list:
    - key: 'ssh-rsa AAAAB3NzaC1yc...'
      user: 'drybjed'

    - key: 'ssh-rsa AAAAB3NzaC1yc...'
      user: 'drybjed'
      options: 'no-port-forwarding,no-agent-forwarding,command="ls /home/drybjed"'

    - key: 'ssh-rsa AAAAB3NzaC1yc...'
      user: 'drybjed'
      state: 'absent'
```

**Comment by jctanner** [comment]
@drybjed a fixed patch:

```
diff --git a/library/system/authorized_key b/library/system/authorized_key
index 7626a9a..47eb2c4 100644
--- a/library/system/authorized_key
+++ b/library/system/authorized_key
@@ -330,10 +330,11 @@ def enforce_state(module, params):

     # Check our new keys, if any of them exist we'll continue.
     for new_key in key:
+        parsed_new_key = parsekey(module, new_key)
         if key_options is not None:
-            new_key = "%s %s" % (key_options, new_key)
+            parsed_options = parseoptions(module, key_options)
+            parsed_new_key = (parsed_new_key[0], parsed_new_key[1], parsed_options, parsed_new_key[3])

-        parsed_new_key = parsekey(module, new_key)
         if not parsed_new_key:
             module.fail_json(msg="invalid key specified: %s" % new_key)
```

**Comment by jurgenhaas** [comment]
@jctanner sorry I gave credit to the wrong person in my previous comment. Of course I do thank _you_ for helping us fixing this.

I have applied this second patch in addition to the first one but it still doesn't resolve the {{item}} token in the key options as described before. Probably this patch wasn't meant to address this issue in the first place.

**Comment by drybjed** [comment]
@jctanner I applied your patch, but nothing changed in regards of my earlier example. `key_options` with `no-port-forwarding,no-agent-forwarding` isn't even added by Ansible to the `authorized_keys` file.

**Comment by jctanner** [comment]
@drybjed I'm not sure why you aren't seeing any options set. With the last patch I always get the options set in the key (albeit in a different order):

```
jtanner@u1304:~$ cat /tmp/nobody/.ssh/testkey.txt 
ssh-rsa XXX_AAA... 
ssh-rsa XXX_NOOPTS... 
```

```
jtanner@u1304:~/issues/5486-authorized_key_options-ignored$ cat createopts.yml 
- hosts: localhost
  gather_facts: False
  vars:
    sshd_authorized_keys_list:
        - key: 'ssh-rsa XXX_OPTS...'
          user: 'nobody'
          options: 'no-port-forwarding,no-agent-forwarding,command="ls /home/nobody"'
          keyfile: /tmp/nobody/.ssh/testkey.txt
  tasks:
    - name: Configure authorized keys for users
      authorized_key: key='{{ item.key }}' 
                      key_options='{{ item.options | default("") }}'
                      path={{ item.keyfile }} 
                      user={{ item.user }}
                      state={{ item.state | default('present') }}
      with_flattened:
        - sshd_authorized_keys_list

    - shell: chown -R jtanner:jtanner /tmp/nobody   
```

```
jtanner@u1304:~/issues/5486-authorized_key_options-ignored$ ansible-playbook -i inventory createopts.yml 

PLAY [localhost] ************************************************************** 

TASK: [Configure authorized keys for users] *********************************** 
changed: [localhost] => (item={'keyfile': '/tmp/nobody/.ssh/testkey.txt', 'user': 'nobody', 'key': 'ssh-rsa XXX_OPTS...', 'options': 'no-port-forwarding,no-agent-forwarding,command="ls /home/nobody"'})

TASK: [shell chown -R jtanner:jtanner /tmp/nobody] **************************** 
changed: [localhost]

PLAY RECAP ******************************************************************** 
localhost                  : ok=2    changed=2    unreachable=0    failed=0  
```

```
jtanner@u1304:~$ cat /tmp/nobody/.ssh/testkey.txt 
ssh-rsa XXX_AAA... 
command="ls /home/nobody",no-agent-forwarding,no-port-forwarding ssh-rsa XXX_OPTS... 
ssh-rsa XXX_NOOPTS... 
```

**Comment by jctanner** [comment]
@jurgenhaas the item var expansion is not handled by the authorized key module. "key_options" was not available in 1.3.4, so I don't think "This used to be working prior to version 1.4 final but is no longer working since." is a valid statement.

```
$ git checkout remotes/origin/release1.3.4
$ ansible-playbook -i inventory createopts.yml 

PLAY [localhost] ************************************************************** 

TASK: [Configure authorized keys for users] *********************************** 
failed: [localhost] => (item={'keyfile': '/tmp/nobody/.ssh/testkey.txt', 'user': 'nobody', 'key': 'ssh-rsa XXX_OPTS...', 'options': 'no-port-forwarding,no-agent-forwarding,command="ls /home/nobody"'}) => {"failed": true, "item": {"key": "ssh-rsa XXX_OPTS...", "keyfile": "/tmp/nobody/.ssh/testkey.txt", "options": "no-port-forwarding,no-agent-forwarding,command=\"ls /home/nobody\"", "user": "nobody"}}
msg: unsupported parameter for module: key_options

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/jtanner/createopts.retry

localhost                  : ok=0    changed=0    unreachable=0    failed=1 
```

**Comment by jurgenhaas** [comment]
@jctanner I can't tell when this used to be working, until very recently we always worked with the latest dev version and I think it was 1.4.1 when we switched to the published release because we went into production and wanted to rely on a stable release.

So, what I'm saying is that we did not use that with version 1.3.4 but with a 1.4 development release. OK, that was on our own risk probably but I don't understand why variables shouldn't be extended in this context as they do everywhere else. And certainly they did in the dev release.
